bft_interp/src/lib.rs

```diff
@@ -1,39 +1,510 @@
+
+        let mut vm: VirtualMachine<u8> = VirtualMachine::new(&test_program, None, true);
+
+        let result = vm.move_head_left();
+        let expected_error: Result<(), VMError> = Err(VMError::HeadUnderrun(bad_instruction));
+
+        assert!(result.is_err());
+        assert_eq!(result, expected_error);
+    }
+
+    // Does moving the head left at the start of an fixed tape error correctly?
+    #[test]
+    fn test_move_head_left_fixed_bad() {
+        let test_program = test_program();
+        let bad_instruction = test_program.instructions()[0].clone();
+
+        let mut vm: VirtualMachine<u8> = VirtualMachine::new(&test_program, None, false);
+
+        let result = vm.move_head_left();
+        let expected_error: Result<(), VMError> = Err(VMError::HeadUnderrun(bad_instruction));
+
+        assert!(result.is_err());
+        assert_eq!(result, expected_error);
+    }
+
+    // Does moving the head right on an extensible tape work when the head has space to move?
+    #[test]
+    fn test_move_head_right_extensible_good() {
+        let test_program = test_program();
+        let tape_len = Some(NonZeroUsize::new(1000).unwrap());
+        let mut vm: VirtualMachine<u8> = VirtualMachine::new(&test_program, tape_len, true);
+
+        let result = vm.move_head_right();
+
+        assert!(result.is_ok());
+        assert_eq!(vm.head, 1);
+    }
+
+    // Does moving the head right on an fixed tape work?
+    #[test]
+    fn test_move_head_right_fixed_good() {
+        let test_program = test_program();
+        let mut vm: VirtualMachine<u8> = VirtualMachine::new(&test_program, None, false);
+
+        let result = vm.move_head_right();
+
+        assert!(result.is_ok());
+        assert_eq!(vm.head, 1);
+    }
+
+    // Does moving the head right at the end of a fixed tape error correctly?
+    #[test]
+    fn test_move_head_right_fixed_bad() {
+        let test_program = test_program();
+        let bad_instruction = test_program.instructions()[0].clone();
+        let tape_len = Some(NonZeroUsize::new(1000).unwrap());
+        let mut vm: VirtualMachine<u8> = VirtualMachine::new(&test_program, tape_len, false);
+        vm.head = 999;
+
+        let result = vm.move_head_right();
+        let expected_error: Result<(), VMError> = Err(VMError::HeadOverrun(bad_instruction));
+
+        assert!(result.is_err());
+        assert_eq!(result, expected_error);
+    }
+
+    // Does moving the head right at the end of an extensible tape make the tape grow?
+    #[test]
+    fn test_auto_tape_extension() {
+        let test_program = test_program();
+        let tape_len = Some(NonZeroUsize::new(1000).unwrap());
+        let mut vm: VirtualMachine<u8> = VirtualMachine::new(&test_program, tape_len, true);
+
+        vm.head = 999;
+
+        let result = vm.move_head_right();
+
+        assert!(result.is_ok());
+        assert_eq!(vm.head, 1000);
+        assert_eq!(vm.cells.capacity(), 2000)
```

Comment:
I would be looking for `vm.cells.len()` equal to `1001` not anything to do with capacity which is just about the underlying memory allocation.